### PR TITLE
GH-117536: GH-117894: fix athrow().throw(...) unawaited warning

### PR DESCRIPTION
**Pull request by graingert**
* Issue: gh-117536

* Issue: gh-117894

### CODE REVIEW

**Comment by graingert** [comment]
This also fixes a bug where an athrow()/aclose() can be reused across StopIteration calls, so needs backporting to 3.12

**Comment by graingert** [comment]
Here's the repro:

```
>>> class MyExc(Exception): pass
...
>>> import itertools
>>> async def gen():
...     for i in itertools.count():
...         try:
...             yield i
...         except MyExc:
...             pass
...
>>> agen = gen()
>>> agen.asend(None).send(None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
StopIteration: 0
>>> athrow = agen.athrow(MyExc)
>>> athrow.throw(MyExc)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
StopIteration: 1
>>> athrow.throw(MyExc)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
StopIteration: 2
>>> athrow.throw(MyExc)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
StopIteration: 3
>>> athrow.throw(MyExc)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
StopIteration: 4
>>>
```

File: Misc/NEWS.d/next/Core and Builtins/2024-04-13-16-55-53.gh-issue-117536.xkVbfv.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix a :exc:`RuntimeWarning` when calling ``agen.aclose().throw(Exception)``
```
Comment (graingert) [inline Misc/NEWS.d/next/Core and Builtins/2024-04-13-16-55-53.gh-issue-117536.xkVbfv.rst]:
```suggestion
Fix a :exc:`RuntimeWarning` when calling ``agen.aclose().throw(Exception)``
Prevent ``agen.aclose()`` objects being re-used after ``.throw()``
```

File: Misc/NEWS.d/next/Core and Builtins/2024-04-13-16-55-53.gh-issue-117536.xkVbfv.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
...
+Prevent ``agen.aclose()`` objects being re-used after ``.throw()``
```
Comment (graingert) [inline Misc/NEWS.d/next/Core and Builtins/2024-04-13-16-55-53.gh-issue-117536.xkVbfv.rst]:
```suggestion
```

Comment (vstinner) [inline Misc/NEWS.d/next/Core and Builtins/2024-04-13-16-55-53.gh-issue-117536.xkVbfv.rst]:
```suggestion
Fix a :exc:`RuntimeWarning` when calling ``agen.aclose().throw(Exception)``.
Prevent ``agen.aclose()`` objects being re-used after ``.throw()``.
```

Comment (graingert) [reply Misc/NEWS.d/next/Core and Builtins/2024-04-13-16-55-53.gh-issue-117536.xkVbfv.rst]: Ah I split this across two blurbs now so this is out of date

File: Misc/NEWS.d/next/Core and Builtins/2024-04-13-16-55-53.gh-issue-117536.xkVbfv.rst (hunk 1 above)
Comment (graingert) [inline Misc/NEWS.d/next/Core and Builtins/2024-04-13-16-55-53.gh-issue-117536.xkVbfv.rst]:
```suggestion
Fix a :exc:`RuntimeWarning` when calling ``agen.aclose().throw(Exception)``.
```

File: Misc/NEWS.d/next/Core and Builtins/2024-04-15-13-53-59.gh-issue-117894.8LpZ6m.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+Prevent ``agen.aclose()`` objects being re-used after ``.throw()``
```
Comment (graingert) [inline Misc/NEWS.d/next/Core and Builtins/2024-04-15-13-53-59.gh-issue-117894.8LpZ6m.rst]:
```suggestion
Prevent ``agen.aclose()`` objects being re-used after ``.throw()``.
```

File: Lib/test/test_asyncgen.py [hunk 4]
```diff
@@ -1671,6 +1659,42 @@ async def run():
...
+        class MyException(Exception):
+            pass
+
+        nxt = it.aclose()
+        with self.assertRaises(MyException):
+            nxt.throw(MyException)
+
+        with self.assertRaisesRegex(
+            RuntimeError,
+            r"cannot reuse already awaited aclose\(\)/athrow\(\)"
```
Comment (agronholm) [inline Lib/test/test_asyncgen.py:1696]: Should `athrow()` also be tested if it's explicitly mentioned here?

File: Lib/test/test_asyncgen.py [hunk 5]
```diff
@@ -1752,6 +1776,17 @@ async def gen():
             g.aclose()
             gc_collect()
 
+    def test_aclose_throw(self):
```
Comment (agronholm) [inline Lib/test/test_asyncgen.py:1801]: This test seems to pass even without your patch.

File: Lib/test/test_asyncgen.py [hunk 6]
```diff
@@ -399,11 +399,8 @@ async def gen():
 
         with self.assertWarns(DeprecationWarning):
             x = gen().athrow(GeneratorExit, GeneratorExit(), None)
-        with self.assertWarnsRegex(RuntimeWarning,
-                f"coroutine method 'athrow' of '{gen.__qualname__}' "
-                f"was never awaited"):
-            del x
-            gc_collect()
+        with self.assertRaises(GeneratorExit):
+            x.send(None)
```
Comment (agronholm) [inline Lib/test/test_asyncgen.py]: This test seems to pass even without the patch. What is the `x.send(None)` for here?

File: Lib/test/test_asyncgen.py (hunk 5 above)
Comment (graingert) [reply Lib/test/test_asyncgen.py:1801]: Ah this might need a gc.collect()

File: Lib/test/test_asyncgen.py (hunk 6 above)
Comment (graingert) [reply Lib/test/test_asyncgen.py]: This might also need a gc.collect()

Comment (agronholm) [reply Lib/test/test_asyncgen.py]: It's still passing on vanilla 3.12? What is the expected result without your patch?

File: Lib/test/test_asyncgen.py (hunk 5 above)
Comment (agronholm) [reply Lib/test/test_asyncgen.py:1801]: This still passes too.

File: Lib/test/test_asyncgen.py [hunk 7]
```diff
@@ -399,11 +399,9 @@ async def gen():
...
+            gc.collect()
```
Comment (graingert) [inline Lib/test/test_asyncgen.py]:
```suggestion
            del x
            gc.collect()
```

File: Lib/test/test_asyncgen.py [hunk 8]
```diff
@@ -1752,6 +1797,18 @@ async def gen():
...
+        async def gen():
+            return
+            yield
+
+        class MyException(Exception):
+            pass
+
+        g = gen()
+        with self.assertRaises(MyException):
+            g.aclose().throw(MyException)
```
Comment (graingert) [inline Lib/test/test_asyncgen.py:1811]:
```suggestion
            g.aclose().throw(MyException)
            del g
```

File: Lib/test/test_asyncgen.py [hunk 9]
```diff
@@ -399,11 +399,10 @@ async def gen():
...
         with self.assertWarns(DeprecationWarning):
             x = gen().athrow(GeneratorExit, GeneratorExit(), None)
-        with self.assertWarnsRegex(RuntimeWarning,
-                f"coroutine method 'athrow' of '{gen.__qualname__}' "
-                f"was never awaited"):
+        with self.assertRaises(GeneratorExit):
+            x.send(None)
             del x
-            gc_collect()
+            gc.collect()
```
Comment (graingert) [inline Lib/test/test_asyncgen.py]:
```suggestion
            gc_collect()
```

File: Lib/test/test_asyncgen.py [hunk 10]
```diff
@@ -1752,6 +1798,19 @@ async def gen():
...
+            yield
+
+        class MyException(Exception):
+            pass
+
+        g = gen()
+        with self.assertRaises(MyException):
+            g.aclose().throw(MyException)
+            del g
+            gc.collect()
```
Comment (graingert) [inline Lib/test/test_asyncgen.py]:
```suggestion
            gc_collect()
```

**Review by agronholm — APPROVED** [review]
LGTM.

**Review by encukou — APPROVED** [review]
I'm not an expert in this area, but I cannot find anything wrong with the patch. An unhandled exception marking the awaitable closed (awaited) makes sense.

**Comment by graingert** [comment]
@encukou can I also get a review on https://github.com/python/cpython/pull/117882 ?

**Comment by encukou** [comment]
Let's merge this to fix the blocker issue. On #117882 I'll try to deepen my understanding of this code, so I can give a better review.

**Comment by graingert** [comment]
This needs backporting to 3.12

**Pull request merged**